presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -548,6 +548,13 @@ public PlanWithProperties visitTableWriter(TableWriterNode node, Context context
             PlanWithProperties source = node.getSource().accept(this, context);
 
             Optional<PartitioningScheme> partitioningScheme = node.getPartitioningScheme();
+
+            if (partitioningScheme.isPresent() && source.getProperties().getNodePartitioning().isPresent() &&
+                    partitioningScheme.get().getPartitioning().equals(source.getProperties().getNodePartitioning().get())) {
+                // if the source is already partitioned as required, just use it
+                return rebaseAndDeriveProperties(node, source);
+            }
```

Comment:
This appears very complicated. This `if` and all code below (with this `if` and the next `if` swapped) essentially boils down to:

```
  if (!partitioningScheme.isPresent && redistributedWrites) {
    partitioningScheme = Optional.of(...);
  }
  
  if (partitioningScheme.isPresent()) {
    source = add exchange to source so that it matches partitioningScheme (if not already)
  }

  return rebaseAndDeriveProperties(node, source);
```

As a result, I find this code extremely complicated. Is there a better way to write this, @martint?

Reply:
An alternatively way would be just to write the final `if` as

```java
            if (partitioningScheme.isPresent() && !(source.getProperties().getNodePartitioning().isPresent() && partitioningScheme.get().getPartitioning().equals(source.getProperties().getNodePartitioning().get()))) {
                source = withDerivedProperties(
                        partitionedExchange(
                                idAllocator.getNextId(),
                                REMOTE,
                                source.getNode(),
                                partitioningScheme.get()),
                        source.getProperties());
            }
            else {
                // do nothing and use the original source
            }

```

Reply:
Or we can add some method like `requiresPartitionedExchange(PlanWithProperties, Optional<PartitioningScheme>)` and do something like

```
if  (!requiresPartitionedExchange(source, partitioningScheme)) {
   source = withDerivedProperties(...)
}
```